### PR TITLE
feat(bigtable): Add support for reverse scans

### PR DESCRIPTION
**Pull request by brandtnewton**

### CODE REVIEW

File: bigtable/bigtable.go [hunk 1]
```diff
@@ -372,13 +440,27 @@ func (r RowRange) String() string {
...
 
 func (r RowRange) proto() *btpb.RowSet {
-	rr := &btpb.RowRange{
-		StartKey: &btpb.RowRange_StartKeyClosed{StartKeyClosed: []byte(r.start)},
+	var rr = btpb.RowRange{}
+
+	// empty strings are NOT considered unbounded for backwards compatibility
+	if r.startBound == unbounded {
+		// leave unbounded
+	} else if r.startBound == closed || r.startBound == undefined {
```
Comment (igorbernstein2) [inline bigtable/bigtable.go:448]: can we normalize these during construction to avoid dealing with ambiguity ?

File: bigtable/bigtable.go [hunk 2]
```diff
@@ -450,8 +555,18 @@ func PrefixRange(prefix string) RowRange {
...
+
+// InfiniteReverseRange returns the RowRange consisting of all keys at least as
+// large as start.
+func InfiniteReverseRange(limit string) RowRange {
+	return RowRange{
+		limit:      limit,
+		startBound: unbounded,
+		endBound:   closed,
 	}
 }
```
Comment (igorbernstein2) [inline bigtable/bigtable.go:571]: can we group all of the constructors together?

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: bigtable/bigtable.go [hunk 3]
```diff
@@ -331,70 +346,206 @@ func (r RowList) retainRowsAfter(lastRowKey string) RowSet {
...
+		start:      start,
+		endBound:   endBound,
+		end:        end,
 	}
 }
 
 // Unbounded tests whether a RowRange is unbounded.
 func (r RowRange) Unbounded() bool {
-	return r.limit == ""
+	return r.startBound == rangeUnbounded || r.endBound == rangeUnbounded
```
Comment (brandtnewton) [inline bigtable/bigtable.go:417]: technically, this is NOT backwards compatible.

File: bigtable/bigtable.go [hunk 4]
```diff
@@ -331,70 +346,248 @@ func (r RowList) retainRowsAfter(lastRowKey string) RowSet {
...
+		return true
+	}
+
+	if r.startBound == rangeOpen || r.endBound == rangeOpen {
+		return r.start < r.end
+	} else if r.startBound == rangeClosed {
+		return r.start <= r.end
+	} else {
+		return true
+	}
```
Comment (igorbernstein2) [inline bigtable/bigtable.go:590]: This pretty hard to reason about and is not immediately clear why this is correct. I would either heavily comment this or make it more explicit. ie

```
// if either end is unbounded, then the range is always valid
if r.Unbounded() {
   return true
}

// If either end is an open interval, then the start must be strictly less than the end
// And since neither end is unbounded, we don't have to check for empty strings
if r.startBound == rangeOpen || r.endBound == rangeOpen {
  return r.start < r.end
}

// At this point both endpoints must be closed, which makes [a,a] a valid interval
return r.start <= r.end
```

File: bigtable/integration_test.go [hunk 5]
```diff
@@ -885,7 +918,7 @@ func TestIntegration_Read(t *testing.T) {
 	}{
 		{
 			desc: "read all, unfiltered",
-			rr:   RowRange{},
```
Comment (igorbernstein2) [inline bigtable/integration_test.go:888]: I'm concerned that this had to change....it implies a breaking change. I think the root of the problem is that rangeUnbounded is not mapped to 0 in your enum

File: bigtable/internal/testproxy/proxy.go [hunk 6]
```diff
@@ -573,6 +573,12 @@ func (s *goTestProxyServer) ReadRows(ctx context.Context, req *pb.ReadRowsReques
 		rs = bigtable.InfiniteRange("")
 	}
 
+	var opts []bigtable.ReadOption
```
Comment (igorbernstein2) [inline bigtable/internal/testproxy/proxy.go:576]: Please move the proxy changes to a separate PR

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: bigtable/bigtable.go (hunk 4 above)
Comment (brandtnewton) [reply bigtable/bigtable.go]: Agreed. Thanks for the suggestion!

File: bigtable/bigtable.go (hunk 3 above)
Comment (brandtnewton) [reply bigtable/bigtable.go:449]: Previously:
```
rr := NewRange("", "z")
rr.String() // ["", "z")
rr.Unbounded() // false
```
but now:
```
rr := NewRange("", "z")
rr.String() // (∞, "z")
rr.Unbounded() // true
```
Now I realize that String() is also not backwards compatible this way

File: bigtable/integration_test.go (hunk 5 above)
Comment (brandtnewton) [reply bigtable/integration_test.go]: True. I think changing the default rangeBound to rangeUnbounded will preserve the behavior of RowRange better - although direct use of the RowRange constructor is discouraged

**Review by igorbernstein2 — APPROVED** [review]

File: bigtable/integration_test.go (hunk 5 above)
Comment (igorbernstein2) [reply bigtable/integration_test.go]: Agreed, but its almost definitely being used by endusers.
